File: rust/src/bin/rpc_client.rs
use futures::StreamExt;
use lapin::{
    options::*, types::FieldTable, types::ShortString, BasicProperties, Channel, Connection,
    ConnectionProperties, Consumer, Queue,
};
use std::convert::TryInto;
use std::fmt::Display;
use uuid::Uuid;

#[derive(Debug)]
enum Error {
    CannotDecodeReply,
    NoReply,
}

impl std::error::Error for Error {}

impl Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            Error::CannotDecodeReply => write!(f, "Cannot decode reply"),
            Error::NoReply => write!(f, "No reply arrived"),
        }
    }
}

struct FibonacciRpcClient {
    conn: Connection,
    channel: Channel,
    callback_queue: Queue,
    consumer: Consumer,
    correlation_id: ShortString,
}

impl FibonacciRpcClient {
    async fn new() -> Result<Self, lapin::Error> {
        let addr = "amqp://127.0.0.1:5672";
        let conn = Connection::connect(addr, ConnectionProperties::default()).await?;
        let channel = conn.create_channel().await?;
        let callback_queue = channel
            .queue_declare(
                "",
                QueueDeclareOptions {
                    exclusive: true,
                    ..Default::default()
                },
                FieldTable::default(),
            )
            .await?;

        let consumer = channel
            .basic_consume(
                callback_queue.name().as_str(),
                "rpc_client",
                BasicConsumeOptions {
                    no_ack: true,
                    ..Default::default()
                },
                FieldTable::default(),
            )
            .await?;

        let correlation_id = Uuid::new_v4().to_string().into();

        Ok(Self {
            conn,
            channel,
            callback_queue,
            consumer,
            correlation_id,
        })
    }

    async fn call(&mut self, n: u64) -> Result<u64, Box<dyn std::error::Error>> {
        self.channel
            .basic_publish(
                "",
                "rpc_queue",
                BasicPublishOptions::default(),
                &*n.to_le_bytes().to_vec(),
                BasicProperties::default()
                    .with_reply_to(self.callback_queue.name().clone())
                    .with_correlation_id(self.correlation_id.clone()),
            )
            .await?
            .await?;

        while let Some(delivery) = self.consumer.next().await {
            if let Ok(delivery) = delivery {
                if delivery.properties.correlation_id().as_ref() == Some(&self.correlation_id) {
                    return Ok(u64::from_le_bytes(
                        delivery
                            .data
                            .as_slice()
                            .try_into()
                            .map_err(|_| Error::CannotDecodeReply)?,
                    ));
                }
            }
        }

        Err(Box::new(Error::NoReply))
    }

    async fn close(&self) -> Result<(), lapin::Error> {
        self.conn.close(0, "").await
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut fibonacci_rpc = FibonacciRpcClient::new().await?;
    println!(" [x] Requesting fib(30)");
    let response = fibonacci_rpc.call(30).await?;
    println!(" [.] Got {}", response);
    fibonacci_rpc.close().await?;
    Ok(())
}

File: zig/src/rpc_client.zig
const std = @import("std");
const bunny = @import("bunny");

pub fn main(init: std.process.Init) !void {
    const allocator = std.heap.smp_allocator;
    const io = init.io;
    const args = try init.minimal.args.toSlice(init.arena.allocator());

    var stdout_buf: [256]u8 = undefined;
    var stdout_w = std.Io.File.stdout().writerStreaming(io, &stdout_buf);
    const out = &stdout_w.interface;
    defer out.flush() catch {};

    const n: u64 = if (args.len > 1)
        std.fmt.parseInt(u64, args[1], 10) catch 30
    else
        30;

    const conn = try bunny.Connection.open(allocator, .{
        .recovery = .{ .enabled = false },
    });
    defer conn.deinit();

    const ch = try conn.openChannel();
    defer ch.close();

    var reply_queue = try ch.temporaryQueue();
    defer reply_queue.deinit(allocator);
    _ = try reply_queue.subscribe(.automatic);

    const correlation_id = bunny.correlationId(io);

    var req_buf: [32]u8 = undefined;
    const request = try std.fmt.bufPrint(&req_buf, "{d}", .{n});

    try out.print(" [x] Requesting fib({d})\n", .{n});
    try out.flush();

    try ch.publishToQueue("rpc_queue", request, bunny.BasicProperties.default
        .withCorrelationId(&correlation_id)
        .withReplyTo(reply_queue.name));

    while (try ch.recvDelivery()) |delivery| {
        defer delivery.deinit(allocator);

        const cid = delivery.properties.correlation_id orelse continue;
        if (!std.mem.eql(u8, cid, &correlation_id)) continue;

        try out.print(" [.] Got {s}\n", .{delivery.body});
        return;
    }
}
